src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -614,6 +622,9 @@ public bool Remove(TKey key)
                         {
                             entries[last].next = entries[i].next;
                         }
+
+                        value = entries[i].value;
```

Comment:
Duplicating the code might be the only option that doesn't impact performance. But it too has its cost since we're looking at a generic type. Using a branch to avoid the unnecessary copy might be a reasonable compromise.

But before doing any of that the performance impact should be measured.